### PR TITLE
safely find packageable element in pure model for artifact generation

### PR DESCRIPTION
**Pull request by MauricioUyaguari**
- safely find packagable element in pure model
- special casing for functions using function signature
- continue on to next element if unable to find it in pureModel

### CODE REVIEW

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            throw new RuntimeException("Unable to build PureModelContextData with path '" + path + "'", e);
+        }
+    }
+
+    @Test
+    public void testArtifactGenerationFactory()
+    {
+        PureModelContextData data = getPureModelContextDataFromPath("ArtifactGenerationFactoryTestData.json");
+        PureModel pureModel = new PureModel(data, null, DeploymentMode.TEST);
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:53]:
```suggestion
        PureModel pureModel = new PureModel(data, null, DeploymentMode.PROD);
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        ArtifactGenerationFactory factory = new ArtifactGenerationFactory(pureModel, data, data.getElements());
+        Assert.assertEquals(2, factory.getExtensions().size());
+        TestSimpleFunctionArtifactGenerationExtension functionArtifactGenerationExtension = LazyIterate.selectInstancesOf(factory.getExtensions(), TestSimpleFunctionArtifactGenerationExtension.class).getFirst();
+        TestSimpleEnumArtifactGenerationExtension enumArtifactGenerationExtension =  LazyIterate.selectInstancesOf(factory.getExtensions(), TestSimpleEnumArtifactGenerationExtension.class).getFirst();
+        MutableMap<ArtifactGenerationExtension, List<ArtifactGenerationResult>> results = factory.generate();
+        Assert.assertEquals(2, results.size());
+        // check enum generation extension
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:60]: There's so much in this test that it's hard to read. Let's add some spacing to separate different parts out.
```suggestion
        Assert.assertEquals(2, results.size());
        
        // check enum generation extension
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        Assert.assertEquals(1, enumFullResults.size());
+        ArtifactGenerationResult enumResult = enumFullResults.get(0);
+        Assert.assertEquals(TestSimpleEnumArtifactGenerationExtension.class, enumResult.getGenerator().getClass());
+        Assert.assertEquals(enumResult.getElement(), pureModel.getPackageableElement("model::MyEnum"));
+        Assert.assertEquals(1, enumResult.getResults().size());
+        GenerationOutput enumOutput = enumResult.getResults().get(0);
+        Assert.assertEquals("Some output for enumeration 'MyEnum'", enumOutput.getContent());
+        Assert.assertEquals("txt", enumOutput.getFormat());
+        Assert.assertEquals("SomeTestOutput.txt", enumOutput.getFileName());
+        // check function generation results
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:71]:
```suggestion
        Assert.assertEquals("SomeTestOutput.txt", enumOutput.getFileName());
        
        // check function generation results
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        Assert.assertEquals(TestSimpleFunctionArtifactGenerationExtension.class, functionResult.getGenerator().getClass());
+        Assert.assertEquals(functionResult.getElement(), pureModel.getPackageableElement("model::myFunction_String_1__String_1_"));
+        Assert.assertEquals(1, functionResult.getResults().size());
+        GenerationOutput functionOutput = functionResult.getResults().get(0);
+        Assert.assertEquals("Some output for function 'myFunction_String_1__String_1_'", functionOutput.getContent());
+        Assert.assertEquals("txt", functionOutput.getFormat());
+        Assert.assertEquals("MyFunction.txt", functionOutput.getFileName());
+    }
+
+}
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:84]:
```suggestion
    }
}
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import org.finos.legend.engine.protocol.pure.v1.model.context.PureModelContextData;
+import org.finos.legend.engine.shared.core.deployment.DeploymentMode;
+import org.finos.legend.sdlc.generation.file.GenerationOutput;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestArtifactGenerationFactory
+{
+
+    private PureModelContextData getPureModelContextDataFromPath(String path)
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:35]:
```suggestion
{
    private PureModelContextData getPureModelContextDataFromPath(String path)
```

File: legend-sdlc-generation-file/src/main/java/org/finos/legend/sdlc/generation/artifact/ArtifactGenerationFactory.java [hunk 6]
```diff
@@ -67,21 +74,44 @@ public MutableMap<ArtifactGenerationExtension, List<ArtifactGenerationResult>> g
...
+        }
+        if (packageableElement instanceof Function)
+        {
+            Function elementFunc = (Function) packageableElement;
+            String fullPath = pureModel.buildPackageString(packageableElement._package, HelperModelBuilder.getSignature(elementFunc));
+            element = pureModel.getPackageableElement_safe(fullPath);
+        }
+        if (element == null)
+        {
+            LOGGER.info("Unable to find element '" + packageableElement.getPath() + "' in Pure Model");
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/main/java/org/finos/legend/sdlc/generation/artifact/ArtifactGenerationFactory.java:110]: I think we ought to log this at debug level.
```suggestion
            LOGGER.debug("Unable to find element '{}' in Pure Model", packageableElement.getPath());
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleEnumArtifactGenerationExtension.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.finos.legend.engine.protocol.pure.v1.model.context.PureModelContextData;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.PackageableElement;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Enumeration;
+
+import java.util.List;
+
+public class TestSimpleEnumArtifactGenerationExtension implements ArtifactGenerationExtension
+{
+
+    @Override
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleEnumArtifactGenerationExtension.java:30]:
```suggestion
{
    @Override
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleEnumArtifactGenerationExtension.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    public List<Artifact> generate(PackageableElement packageableElement, PureModel pureModel, PureModelContextData pureModelContextData, String s)
+    {
+        String content = "Some output for enumeration '" + packageableElement._name() + "'";
+        Artifact artifact = new Artifact(content, "SomeTestOutput.txt", "txt");
+        return Lists.mutable.with(artifact);
+    }
+
+}
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleEnumArtifactGenerationExtension.java:50]:
```suggestion
    }
}
```

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleFunctionArtifactGenerationExtension.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public List<Artifact> generate(PackageableElement packageableElement, PureModel pureModel, PureModelContextData pureModelContextData, String s)
+    {
+        String content = "Some output for function '" + packageableElement._name() + "'";
+        Artifact artifact = new Artifact(content, "MyFunction.txt", "txt");
+        return Lists.mutable.with(artifact);
+    }
+
+}
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestSimpleFunctionArtifactGenerationExtension.java:48]:
```suggestion
    }
}
```

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java (hunk 4 above)
Comment (MauricioUyaguari) [reply legend-sdlc-generation-file/src/test/java/org/finos/legend/sdlc/generation/artifact/TestArtifactGenerationFactory.java:85]: apologizes wanted them all to be sync but put them all in sync in the wrong format :)

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**